### PR TITLE
Change routing for shielding question

### PR DESCRIPTION
**Pull request by whpearson**
This change uses the information from the shielding advice in the
session to determine what questions should be asked.

If shielding advice is 1 it should ask the basic-care-needs question.

Going to wait for https://github.com/alphagov/govuk-shielded-vulnerable-people-service/pull/251 to address local testing changes needed for e2e.

### CODE REVIEW

File: vulnerable_people_form/form_pages/shared/routing.py [hunk 1]
```diff
@@ -160,10 +161,11 @@ def _get_next_form_url_based_on_location_tier(redirect_url, should_redirect_to_n
 def _get_next_form_url_after_shopping_and_priority_supermarket():
     if _is_tiering_logic_enabled():
         location_tier = get_location_tier()
+        shielding_advice = get_shielding_advice()
 
         _validate_location_tier_is_at_least_very_high(location_tier)
 
-        if location_tier == PostcodeTier.VERY_HIGH_PLUS_SHIELDING.value:
+        if shielding_advice == 1:
```
Comment (sanisc) [inline vulnerable_people_form/form_pages/shared/routing.py:168]: is it worth creating an enum for this? As the 1 doesn't really mean anything who is unfamiliar with the code base

Comment (whpearson) [reply vulnerable_people_form/form_pages/shared/routing.py:168]: Sure.

File: vulnerable_people_form/form_pages/shared/location_tier.py [hunk 2]
```diff
@@ -25,9 +25,9 @@ def update_location_tier_by_uprn(uprn, app):
 
         lad_code = get_ladcode_from_uprn(uprn)
         if app.shielding_advice.is_la_shielding(lad_code):
-            set_shielding_advice(1)
+            set_shielding_advice(ShieldingAdvice.ADVISED_TO_SHIELD)
```
Comment (sanisc) [inline vulnerable_people_form/form_pages/shared/location_tier.py:28]: probably worth explicitly doing .value and for all the others

**Comment by whpearson** [comment]
We would need to update the values for POSTCODE_TIER_OVERRIDE in concourse before this can be deployed.

File: vulnerable_people_form/form_pages/shared/location_tier.py [hunk 3]
```diff
@@ -47,12 +50,14 @@ def update_location_tier_by_postcode(postcode, app):
 
         lad_code = get_ladcode_from_postcode(postcode)
         if app.shielding_advice.is_la_shielding(lad_code):
-            set_shielding_advice(1)
+            set_shielding_advice(ShieldingAdvice.ADVISED_TO_SHIELD.value)
         else:
-            set_shielding_advice(0)
+            set_shielding_advice(ShieldingAdvice.NOT_ADVISED_TO_SHIELD.value)
 
         if postcode in app.config["POSTCODE_TIER_OVERRIDE"]:
```
Comment (sanisc) [inline vulnerable_people_form/form_pages/shared/location_tier.py:57]: can this be set on the app it self in init.py and not reference the config property every-time as this logic is in quite a few places

**Review by sanisc — CHANGES_REQUESTED** [review]

File: vulnerable_people_form/form_pages/shared/location_tier.py (hunk 3 above)
Comment (whpearson) [reply vulnerable_people_form/form_pages/shared/location_tier.py]: Done

File: behave/features/e2e_test_features/01.simple_web_submission_entry.feature [hunk 4]
```diff
@@ -18,7 +18,7 @@ Feature: COVID-19 Shielded vulnerable people service - basic e2e user journey -
 
     Scenario: Should be re-directed to shielding because vulnerable when eligible postcode entered
```
Comment (sanisc) [inline behave/features/e2e_test_features/01.simple_web_submission_entry.feature:19]: do we have any tests where shielding is false e.g for tier 4 or 3?

File: vulnerable_people_form/form_pages/shared/location_tier.py [hunk 5]
```diff
@@ -30,12 +30,15 @@ def update_location_tier_by_uprn(uprn, app):
 
         lad_code = get_ladcode_from_uprn(uprn)
         if app.shielding_advice.is_la_shielding(lad_code):
-            set_shielding_advice(1)
+            set_shielding_advice(ShieldingAdvice.ADVISED_TO_SHIELD.value)
```
Comment (sanisc) [inline vulnerable_people_form/form_pages/shared/location_tier.py:33]: This method seems to do more than just update location tier it sets the shielding advice? Should this be broken down into a few methods:

one to set the tier
one to set the shielding advice
one to set the override for the postcode

?

File: vulnerable_people_form/form_pages/shared/routing.py [hunk 6]
```diff
@@ -158,10 +160,11 @@ def _get_next_form_url_based_on_location_tier(redirect_url, should_redirect_to_n
 def _get_next_form_url_after_shopping_and_priority_supermarket():
     if _is_tiering_logic_enabled():
         location_tier = get_location_tier()
+        shielding_advice = get_shielding_advice()
```
Comment (sanisc) [inline vulnerable_people_form/form_pages/shared/routing.py:163]: This method needs updating `get_redirect_for_returning_user_based_on_tier` as it currently doesn't check the shielding advice - this is for an nhs login user returning back to their submission as well as some one on check answers clicking the change address

**Review by sanisc — CHANGES_REQUESTED** [review]

File: behave/features/e2e_test_features/01.simple_web_submission_entry.feature (hunk 4 above)
Comment (whpearson) [reply behave/features/e2e_test_features/01.simple_web_submission_entry.feature:19]: All the current tier 3 tests are for shielding false.

File: vulnerable_people_form/form_pages/shared/location_tier.py (hunk 5 above)
Comment (whpearson) [reply vulnerable_people_form/form_pages/shared/location_tier.py]: I've changed the name of the function to represent what it does better and it now calls the different functions you suggested,

File: vulnerable_people_form/form_pages/check_your_answers.py [hunk 7]
```diff
@@ -16,7 +16,7 @@ def get_check_your_answers():
     exclude_answers = None
 
     if current_app.is_tiering_logic_enabled:
-        if is_very_high_plus_shielding_without_basic_care_needs_answer():
+        if is_shielding_without_basic_care_needs_answer():
             return redirect(append_querystring_params("/basic-care-needs"))
         if get_location_tier() == PostcodeTier.VERY_HIGH.value:
```
Comment (sanisc) [inline vulnerable_people_form/form_pages/check_your_answers.py:21]: i think this if statement should check if they are not shielding then exclude basic care needs
`
if get_shielding_advice() == ShieldingAdvice.NOT_ADVISED_TO_SHIELD.value:`

File: vulnerable_people_form/integrations/location_eligibility.py [hunk 8]
```diff
@@ -10,6 +10,16 @@
 _DEFAULT_TIER = 1
 
 
+def get_uprn_shielding(uprn, app):
```
Comment (sanisc) [inline vulnerable_people_form/integrations/location_eligibility.py:13]: maybe rename these to get_shielding_advise_by_uprn and get_shielding_advise_by_postcode?

**Review by sanisc — CHANGES_REQUESTED** [review]

File: vulnerable_people_form/form_pages/address_lookup.py [hunk 9]
```diff
@@ -27,7 +27,7 @@ def get_address_lookup():
         }
         return redirect("/support-address")
     except postcode_lookup_helper.NoAddressesFoundAtPostcode:
-        if postcode in current_app.config["POSTCODE_TIER_OVERRIDE"]:
+        if postcode in current_app.postcode_tier_overrride:
```
Comment (sh31) [inline vulnerable_people_form/form_pages/address_lookup.py:30]: Typo in ove**rrr**ide

Comment (whpearson) [reply vulnerable_people_form/form_pages/address_lookup.py]: Fixed.

File: vulnerable_people_form/form_pages/address_lookup.py [hunk 10]
```diff
@@ -59,7 +59,7 @@ def get_address_lookup():
 
 def _create_test_address(postcode):
     return [{
-        "text": f"{current_app.config['POSTCODE_TIER_OVERRIDE'][postcode]}, Test Lane, City, {postcode}",
+        "text": f"{current_app.postcode_tier_override[postcode]}, Test Lane, City, {postcode}",
```
Comment (sh31) [inline vulnerable_people_form/form_pages/address_lookup.py:62]: `building_and_street_line_1:` needs to be updated to use `current_app.postcode_tier_override[postcode]` too

**Review by sanisc — APPROVED** [review]

**Review by sanisc — DISMISSED** [review]

**Review by sh31 — DISMISSED** [review]

**Review by sanisc — APPROVED** [review]

**Pull request merged**